dev/com.ibm.ws.kernel.service/src/com/ibm/wsspi/kernel/service/utils/AtomicServiceReference.java

```diff
@@ -29,126 +31,130 @@
         }
 
         @Override
         public String toString() {
-            return "ctx=" + context + ",ref=" + serviceRef + ",svc=" + locatedService;
+            return "ref=" + serviceRef + ",svc=" + locatedService;
+        }
+
+        @Override
+        public int compareTo(ReferenceTuple<T> o) {
+            // reverse sort because priority queue peeks the least
+            return -(this.serviceRef.compareTo(o.serviceRef));
+        }
+
+        @SuppressWarnings("rawtypes")
+        @Override
+        public boolean equals(Object o) {
+            if (o instanceof ReferenceTuple) {
+                return this.serviceRef.equals(((ReferenceTuple) o).serviceRef);
+            }
+            return false;
         }
     }
 
-    private final AtomicReference<ReferenceTuple<T>> tuple = new AtomicReference<ReferenceTuple<T>>(new ReferenceTuple<T>(null, null, null));
+    // size of 2; it will be very uncommon to this to be used with more than 2 at a time
+    private final PriorityQueue<ReferenceTuple<T>> references = new PriorityQueue<>(2);
 
     /**
      * Create a new AtomicServiceReference for the named service.
      * e.g. from bnd.bnd: referenceName=.... or from component.xml: <reference name="referenceName".... >
-     * 
+     *
      * @param name Name of DS reference
      */
     public AtomicServiceReference(String name) {
         referenceName = name;
     }
 
     public void activate(ComponentContext context) {
-        ReferenceTuple<T> previous = null;
-        ReferenceTuple<T> newTuple = null;
-
-        // Try this until we either manage to make our update or we know we don't have to
-        do {
-            // Get the current tuple
-            previous = tuple.get();
-
-            // If we don't have a previous value
-            newTuple = new ReferenceTuple<T>(context, previous.serviceRef, null);
-
-            // Try to set our new value 
-        } while (!tuple.compareAndSet(previous, newTuple));
+        contextRef.set(context);
     }
 
     public void deactivate(ComponentContext context) {
-        ReferenceTuple<T> previous = null;
-        ReferenceTuple<T> newTuple = null;
-
-        // Try this until we manage to set/replace the value
-        do {
-            // Get the current tuple
-            previous = tuple.get();
-
-            // Create a new tuple that clears out the context (preserve the reference to the service,
-            // as that can have attributes that should be accessible by a user between deactivate
-            // and unset
-            newTuple = new ReferenceTuple<T>(null, previous.serviceRef, null);
-
-            // Try to save the new tuple: retry if someone changed the value meanwhile
-        } while (!tuple.compareAndSet(previous, newTuple));
+        contextRef.set(null);
+        synchronized (references) {
+            // clear services located from the context that is deactivating
+            references.forEach((r) -> r.locatedService.set(null));
```

Comment:
Why does `deactivate()` not reset references to empty (i.e. all ASR state) rather than reset the located services within references?  There only one context per ASR, and it's reset to null.

Reply:
I kept the previous behavior where it nulled out the `ComponentContext` and the instance with a new ReferenceTuple but preserved the reference.  From the old code:

```java
            // Create a new tuple that clears out the context (preserve the reference to the service,
            // as that can have attributes that should be accessible by a user between deactivate
            // and unset
            newTuple = new ReferenceTuple<T>(null, previous.serviceRef, null);
```